docs/proposals/env-vars.md

```diff
@@ -0,0 +1,563 @@
+}
+```
+
+This proposal is making the simplifying assumption that ultimately setting the `Env []EnvVar` field in the 
+underlying Tekton `Step`/`Container` is sufficient for an initial implementation and scenarios we want to support.
+That will help the initial implementation in that the explicit "name" index field helps with default kubernetes 
+"path" API operations that tend to occur with devops/gitops flows.  Also, `ValueFrom *EnvVarSource` still allows
+us to pull content from `Secrets`, like authentication token.  Many of the tools when building images (including
+source code extraction and tools that have to download dependencies) consume authentication tokens via environment
+variables.  Storing those token in k8s `Secrets` vs. having to specify them in scripts referenced in `Containers` has 
+become a preferred method when building images in k8s native environments.
+
+Next, the question of whether to use the k8s `EnvVar` directly exists.  This proposal suggests creating a new
+Shipwright type that inlines the k8s type, in case we want to extend and add new features later that require
+additional fields or new types from k8s or Tekton.  So with that, we start with
+
+```go
+type EnvironmentVariable struct {
+   corev1.EnvVar
+   
+}
+```
+
+NOTE: `valueFrom` in k8s has a range of options, from `Secret` refs, `ConfigMap` refs, field refs, etc.  The `Secret` ref
+is the one we see immediate need for, but the EP does *NOT* proscribe restrictions at this time.  Of course, as we progress
+toward implementation and customer feedback, if need be, restrictions can be added at a later date.
+
+Then, let's examine the updates for `BuildStrategy` (both cluster and namespaced scope), `Build`, and `BuildRun` and where we
+might update them with an array of `EnvironmentVariable`
+
+`BuildStrategy` needs some further consideration, in that `BuildStep` already inlines `Container`, which allows specification 
+of `EnvVar`.
+
+Those existing `BuildStep` / `Container` environment variables can be the way an administrator could ensure that 
+any environment variable exists and has a given setting.  These would be environment variables that could not be 
+changed by a `Build` or `BuildRun`.
+
+But in the case where an organization wants to provide a default environment variable setting, but allow their `Build` and
+`BuildRun` creators to override the value for that setting (including setting it to the empty string, which in most 
+cases would mean "turning off" the environment variable), the `BuildStrategySpec` would have an array of Shipwright
+`EnvironmentVariable` added to it.
+
+Lastly, so there are two basic mindsets from an admin perspective:
+
+- Allow any environment variable in general, but restrict a few
+- Disallow any environment variable in general, but allow a few
+
+To facilitate both mindsets, we'll add two string arrays to the `BuildStrategySpec`
+- `AllowedEnvironmentVariables`
+- `DisallowedEnvironmentVariables`
```

Comment:
This is perhaps a bit too misleading, since this controls env var overrides at the Build/BuildRun level. Perhaps rename as:

1. `allowedEnvVarOverrides`
2. `blockedEnvVarOverrides`

We should validate that only one of the two are set for a given BuildStrategy/ClusterBuildStrategy. This can be done directly on the generated CRD via OpenAPI - no need for validating admission webhooks.

Reply:
No those are suppose to be general allow / block lists.  Though initial reaction I do like `blocked...` better than `disallowed...`

Prevention of overrides at the moment is specifying k/v pairs in the existing BuildStep/k8s Container env var array vs. the new array (which in response to another comment you had, minimally needs a rename, if we keep at all).

At this point, it perhaps makes sense to have a common voice to voice / face to face conversation with the current set of reviewers on the "administrative control" themed features.

Based on where we land with that, we can resume this thread as needed.  Where even if we keep these, field names and the surrounding text in the EP perhaps need to be clearer.

Reply:
> webhook

on the no need for validating webhooks, we may need to cross reference your take here with @SaschaSchwarze0 's https://github.com/shipwright-io/build/pull/726#discussion_r611570295
where part of that I think arose in the recent parameters EP